certbot/compat/filesystem.py

```diff
@@ -0,0 +1,125 @@
+"""Compat module to handle files security on Windows and Linux"""
+from __future__ import absolute_import
+
+import os  # pylint: disable=os-module-forbidden
+import stat
+
+try:
+    import ntsecuritycon  # pylint: disable=import-error
+except ImportError:  # pragma: no cover
+    ntsecuritycon = None  # type: ignore
+try:
+    import win32security  # pylint: disable=import-error
+except ImportError:  # pragma: no cover
+    win32security = None  # type: ignore
+
+
+def chmod(file_path, mode):
+    # type: (str, int) -> None
+    """
+    Apply a POSIX mode on given file_path:
+        * for Linux, the POSIX mode will be directly applied using chmod,
+        * for Windows, the POSIX mode will be translated into a Windows DACL that make sense for
+          Certbot context, and applied to the file using kernel calls.
+    :param str file_path: Path of the file
+    :param int mode: POSIX mode to apply
+    """
+    if not win32security:
+        os.chmod(file_path, mode)
+    else:
+        _apply_win_mode(file_path, mode)
+
+
+def _apply_win_mode(file_path, mode):
+    # Resolve symbolic links
+    if os.path.islink(file_path):
+        link_path = file_path
+        file_path = os.readlink(file_path)
+        if not os.path.isabs(file_path):
+            file_path = os.path.join(os.path.dirname(link_path), file_path)
+    # Get owner sid of the file
+    security = win32security.GetFileSecurity(file_path, win32security.OWNER_SECURITY_INFORMATION)
+    user = security.GetSecurityDescriptorOwner()
+
+    # New DACL, that will overwrite existing one (including inherited permissions)
+    dacl = _generate_dacl(user, mode)
+
+    # Apply the new DACL
+    security.SetSecurityDescriptorDacl(1, dacl, 0)
+    win32security.SetFileSecurity(file_path, win32security.DACL_SECURITY_INFORMATION, security)
+
+
+def _generate_dacl(user_sid, mode):
+    analysis = _analyze_mode(mode)
+
+    # Get standard accounts sid
+    system = win32security.ConvertStringSidToSid('S-1-5-18')
+    admins = win32security.ConvertStringSidToSid('S-1-5-32-544')
+    everyone = win32security.ConvertStringSidToSid('S-1-1-0')
+
+    # New dacl, without inherited permissions
+    dacl = win32security.ACL()
+
+    # If user is already system or admins, any ACE defined here would be superseeded by
+    # the full control ACE that will be added after.
+    if str(user_sid) not in [str(system), str(admins)]:
```

Comment:
What we are testing here is a), not b). The precise meaning here is to not add two times the exactly same ACL.

About your comment on calculating the subset, you just have been triggered doing a "do not repeat yourself" reasoning on Windows ^^

Yes, this would work, to just define in a ACL the missing rights from a wider scoped ACL (on a group). But this is not a Windows style. A file or directory created by Windows will have by default one ACL on the user, on ACL for its primary group, independently of what each ACL contains.

So I prefer to stick with the way to do in Windows and stay consistent towards the platform.

On top of that, I think in general that it is not a good approach to rely on implicit resolution of rights between different scoped ACL. Either:
* you have only a global ACL (for your group) that define your rights
* or you have a specific ACL (for the user) on top of the global one, and then you redefine all the rights on the specific ACL.

Reply:
Indeed. It is not necessarily. I tested (and unit tests confirm) that the `PySID` objects contain a suitable equality operator for comparison. I changed the code to use directly the objects, not their string representations.